src/install.ts

```diff
@@ -0,0 +1,120 @@
+/**
+ * @license
+ * Copyright (c) 2020 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt The complete set of authors may be found
+ * at http://polymer.github.io/AUTHORS.txt The complete set of contributors may
+ * be found at http://polymer.github.io/CONTRIBUTORS.txt Code distributed by
+ * Google as part of the polymer project is also subject to an additional IP
+ * rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+import {promises as fs} from 'fs';
+import path from 'path';
+import {install} from 'pkg-install';
+import pkgUp from 'pkg-up';
+
+export type OnDemandDependencies = Map<string, string>;
+
+export interface ContainsOnDemandDependencies {
+  [index: string]: unknown;
+  devDependencies?: {[index: string]: string};
+  installsOnDemand?: string[];
+}
+
+export const getPackageJSONPath = async(): Promise<string|null> => {
+  return pkgUp({cwd: __dirname});
+};
+
+export const getPackageRoot = async(): Promise<string|null> => {
+  const packageJSONPath = await getPackageJSONPath();
+  return packageJSONPath != null ? path.dirname(packageJSONPath) : null;
+};
+
+/**
+ * Extract a map of allowed "on-demand" dependencies from a given
+ * package.json-shaped object.
+ */
+export const onDemandDependenciesFromPackageJSON =
+    (packageJSON: ContainsOnDemandDependencies): OnDemandDependencies => {
+      const onDemandDependencies = new Map<string, string>();
+
+      const devDependencies = packageJSON?.devDependencies || {};
+      const onDemandList: string[] = packageJSON?.installsOnDemand || [];
+
+      for (const packageName of onDemandList) {
+        if (packageName in devDependencies) {
+          onDemandDependencies.set(packageName, devDependencies[packageName]);
+        }
+      }
+
+      return onDemandDependencies;
+    };
+
+/**
+ * So-called "on-demand" dependencies are any packages that match both of the
+ * following requirements:
+ *
+ *  - They are listed in package.json as "devDependencies"
+ *  - They are enumerated in the non-normative package.json field
+ *    "installsOnDemand"
+ *
+ * This function resolves a map of package names and semver ranges including all
+ * packages that match these requirements.
+ */
+export const getOnDemandDependencies = (() => {
+  let cached: OnDemandDependencies|null = null;
+  return async(): Promise<OnDemandDependencies> => {
+    if (cached == null) {
+      try {
```

Comment:
Can the `try` be scoped more tightly? Shouldn't some/all of the exceptions actually get thrown?

Reply:
Yeah, I think it's just a matter of what UX you want for the module consumer. My thinking here was that if any exceptions are thrown, it may still be possible for Tachometer to run - it just means that we couldn't find a package.json, or read/parse it. We can fail at whatever granularity seems most appropriate to you.